### PR TITLE
Breaks SHA link between NuPIC / NuPIC Core

### LINKED ISSUE #3313: Link NuPIC to public releases of NuPIC Core, not iterative releases
~~Blocked by https://github.com/numenta/nupic.core/issues/1068~~

### PR DESCRIPTION
**Pull request by rhyolight**
Updates Travis-CI and AppVeyor to both get their nupic.bindings versions from requirements.txt, which contains the latest public release version of NuPIC.bindings.

supercedes https://github.com/numenta/nupic/pull/3281

fixes #3313

### CODE REVIEW

**Comment by rhyolight** [comment]
@scottpurdy please review, pretty sure the last build (OS X) will pass.

File: appveyor.yml [hunk 1]
```diff
@@ -61,23 +61,14 @@ install:
...
-  - set BINDINGS_LOC="http://s3-us-west-2.amazonaws.com/artifacts.numenta.org/numenta/nupic.core/releases/nupic.bindings"
-  # Github tagged builds
-  - ps: >-
-      If ($env:APPVEYOR_REPO_TAG -eq "true" -or $env:APPVEYOR_REPO_TAG -eq "True") {
-        pip install $env:BINDINGS_LOC/nupic.bindings-$env:BINDINGS_VERSION-cp27-none-win_amd64.whl
-      }
-      Else {
-        pip install $env:BINDINGS_LOC/nupic.bindings-$env:BINDINGS_VERSION.$env:NUPIC_CORE_COMMITISH-cp27-none-win_amd64.whl
-      }
+  # Install the latest public release stored on S3.
```
Comment (rhyolight) [inline appveyor.yml]: Incorrect comment, will fix.

File: appveyor.yml [hunk 2]
```diff
@@ -115,7 +106,7 @@ after_build:
   - py.test tests\unit
 
   # Python integration tests
-  - py.test tests\integration
+  # - py.test tests\integration
```
Comment (rhyolight) [inline appveyor.yml]: We will be running integration tests in `nupic.regression`. I don't think there is any need to run them here.

File: ci/travis/before_install-linux.sh [hunk 3]
```diff
@@ -43,20 +43,13 @@ python -c 'import pip; print "pip version=", pip.__version__'
...
-# sudo rm -rf /dev/shm && sudo ln -s /run/shm /dev/shm
+while read line; do
+    if [[ $line == nupic.bindings* ]];
+    then
+        IFS='=' read -ra ADDR <<< "$line"
+        NUPIC_BINDINGS_VERSION="${ADDR[2]}"
+    fi
+done <requirements.txt
+
+pip install https://s3-us-west-2.amazonaws.com/artifacts.numenta.org/numenta/nupic.core/releases/nupic.bindings/nupic.bindings-${NUPIC_BINDINGS_VERSION}-cp27-none-linux_x86_64.whl
```
Comment (rhyolight) [inline ci/travis/before_install-linux.sh]: Still have to get the linux binary from S3 until we have manylinux binaries in pypi.

File: ci/travis/before_install-linux.sh [hunk 4]
```diff
@@ -43,20 +43,13 @@ python -c 'import pip; print "pip version=", pip.__version__'
...
-# See: https://github.com/travis-ci/travis-cookbooks/issues/155
-# Commented out to test to see if it works witout it in container mode.
-# sudo rm -rf /dev/shm && sudo ln -s /run/shm /dev/shm
+while read line; do
+    if [[ $line == nupic.bindings* ]];
+    then
+        IFS='=' read -ra ADDR <<< "$line"
+        NUPIC_BINDINGS_VERSION="${ADDR[2]}"
+    fi
+done <requirements.txt
```
Comment (rhyolight) [inline ci/travis/before_install-linux.sh]: This block reads `requirements.txt` and extracts the `nupic.bindings` version.

File: Dockerfile [hunk 5]
```diff
@@ -33,26 +52,19 @@ ENV NTA_DATA_PATH /usr/local/src/nupic/prediction/data
...
-
-WORKDIR /usr/local/src
-
-# Clone nupic.core
-RUN git clone `head -n 2 $NUPIC/.nupic_modules | tail -n 1 | sed -r "s/NUPIC_CORE_REMOTE = '(.+)'/\\1/"` && \
-    cd nupic.core && \
-    git reset --hard `head -n 3 $NUPIC/.nupic_modules | tail -n 1 | sed -r "s/NUPIC_CORE_COMMITISH = '(.+)'/\\1/"`
-
+# Set up nupic.core
+RUN pip install numpy pycapnp
```
Comment (oxtopus) [inline Dockerfile]: There should be parity between the `Dockerfile` implementations in `nupic.core` and this with respect to handling the nupic.bindings dependency. 

In particular, install pip, setuptools, and wheel like this: https://github.com/numenta/nupic.core/blob/master/Dockerfile#L18-L20

Also consider not building nupic.core from source and instead install nupic.bindings from binary wheel.  Recent changes to setup.py allow for correct handling of pycapnp and other dependencies (like numpy), so you shouldn't need `RUN pip install numpy pycapnp` and subsequent lines.

**Comment by oxtopus** [comment]
I like the idea of moving requirements.txt to root, but I recall there being push-back the last time I proposed it.  Maybe bring it up w/ @scottpurdy

**Comment by rhyolight** [comment]
@oxtopus I think he'll approve. He's the one who pushed that commit.

File: Dockerfile (hunk 5 above)
Comment (rhyolight) [inline Dockerfile]: > In particular, install pip, setuptools, and wheel like this: https://github.com/numenta/nupic.core/blob/master/Dockerfile#L18-L20

See above, those three lines are there already: https://github.com/numenta/nupic/pull/3314/files#diff-3254677a7917c6c01f55212f86c57fbfR40



> Also consider not building nupic.core from source and instead install nupic.bindings from binary wheel. Recent changes to setup.py allow for correct handling of pycapnp and other dependencies (like numpy), so you shouldn't need RUN pip install numpy pycapnp and subsequent lines.

I also agree with installing nupic.core from pypi in the dockerfile, but let's do that in another PR.

File: appveyor.yml (hunk 2 above)
Comment (scottpurdy) [inline appveyor.yml]: In general, integration tests are considered part of the primary test suite for CI. I don't care what we do in AV but we should be runnign integration tests for every run in Bamboo.

File: ci/travis/before_install-linux.sh (hunk 4 above)
Comment (scottpurdy) [inline ci/travis/before_install-linux.sh]: Oh you fancy

File: Dockerfile (hunk 5 above)
Comment (scottpurdy) [inline Dockerfile]: It depends on what the Dockerfile is for. Is it a base image for production applications? Or an easy to set up development environment? It's just unclear what the intent is.

**Comment by scottpurdy** [comment]
@oxtopus yeah I moved it to root and much prefer that.

File: ci/travis/script-run-tests.sh [hunk 6]
```diff
@@ -28,7 +28,11 @@ echo
 ${TRAVIS_BUILD_DIR}/scripts/run_nupic_tests.py -u --coverage --failfast || exit
 
 mv ${TRAVIS_BUILD_DIR}/.coverage ${TRAVIS_BUILD_DIR}/.coverage_unit
-# Python integration tests and prep for coveralls reporting
-${TRAVIS_BUILD_DIR}/scripts/run_nupic_tests.py -i --coverage --failfast || exit
 
-mv ${TRAVIS_BUILD_DIR}/.coverage ${TRAVIS_BUILD_DIR}/.coverage_integration
+# Removed because integration tests take so long and we run them in
```
Comment (scottpurdy) [inline ci/travis/script-run-tests.sh]: I'd prefer to revert this change. We shouldn't rely on the regression suite to see if things break. I see it more as testing that we still get good NAB scores and prediction results and for perf testing. Our unit tests aren't sufficient for many breakages so we need integration tests too.

Comment (scottpurdy) [inline ci/travis/script-run-tests.sh]: We should look into the integration tests and make sure they run quickly. But in the meantime, I think longer builds and fewer missed breakages is better than the reverse.

**Comment by scottpurdy** [comment]
Done reviewing.

File: ci/travis/script-run-tests.sh (hunk 6 above)
Comment (rhyolight) [inline ci/travis/script-run-tests.sh]: We removed integration tests from our builds like 2 years ago, mostly because the swarming tests are in there, and they take 20-30 minutes. I'll add them back in, but I think we should try to keep iterative build test phase under 10 minutes.

**Comment by rhyolight** [comment]
@scottpurdy I think we have a compatibility problem because NuPIC is ahead of nupic.core 0.4.5... Standby we might need to cut another release today and stop making changes in NuPIC to reflect Nupic.core changes.

File: Dockerfile (hunk 5 above)
Comment (rhyolight) [inline Dockerfile]: I don't even know.... @oxtopus? @allanino?

**Comment by rhyolight** [comment]
I released nupic.core 0.4.6 and switched this PR to depend on it. Hopefully that clears up compatibility test failures.

**Comment by rhyolight** [comment]
@scottpurdy ready!

**Comment by scottpurdy** [comment]
👍

File: ci/travis/before_install-linux.sh [hunk 7]
```diff
@@ -43,20 +43,13 @@ python -c 'import pip; print "pip version=", pip.__version__'
...
-pip install --no-index --find-links=home/travis/build/numenta/nupic.core/bindings/py/dist/wheels nupic.bindings
-
-# Workaround for multiprocessing.Queue SemLock error from run_opf_bechmarks_test.
-# See: https://github.com/travis-ci/travis-cookbooks/issues/155
-# Commented out to test to see if it works witout it in container mode.
-# sudo rm -rf /dev/shm && sudo ln -s /run/shm /dev/shm
+while read line; do
+    if [[ $line == nupic.bindings* ]];
+    then
+        IFS='=' read -ra ADDR <<< "$line"
```
Comment (vitaly-krugl) [inline ci/travis/before_install-linux.sh]: I have not seen this syntax before: `IFS='='`

Comment (vitaly-krugl) [inline ci/travis/before_install-linux.sh]: Interesting, didn't know IFS was a special variable in bash. Here is a good explanation: http://stackoverflow.com/questions/918886/how-do-i-split-a-string-on-a-delimiter-in-bash

Comment (rhyolight) [inline ci/travis/before_install-linux.sh]: It specifies a delimiter of `=` so each line can be split on `=`.
